### ISSUE TITLE
`helm rollback` to current revision when nothing has changed

### ISSUE THREAD

**Issue by dgn** [istio-ecosystem/sail-operator#429]
We will currently bump the revision number every time we reconcile a resource because we always perform an upgrade. That shouldn't be necessary in most cases. In order to restore the indended state of a helm chart, it is enough to rollback to the current revision.

**Comment by luksa** [comment]
I tried this at some point, but ran into some issues. Can't remember what they were. If we can't get rollback to work, we need to modify the code so that it renders the resources and then we compare them with the current state of those same resources and perform updates only when there's a difference.

**Comment by dgn** [comment]
> I tried this at some point, but ran into some issues. Can't remember what they were. If we can't get rollback to work, we need to modify the code so that it renders the resources and then we compare them with the current state of those same resources and perform updates only when there's a difference.

I thought about that, but wasn't sure comparing all the resources to desired state would be better than current behavior - we're blasting the API server with Get()s at that point. Rolling back to current revision worked in my limited local tests but maybe there's a catch, I do remember us looking at this previously.

**Comment by luksa** [comment]
No, gets are handled locally (they go to the local cache), not the API server.

**Comment by dgn** [comment]
Okay, I tested this again and it turns out that rolling back to a previous revision *also increments the revision*. So there's really not much of a benefit, apart from a more correct helm history (which we're not really keeping anyway, we set `MaxVersion = 1`).

But we can generate the new manifests by doing a dry-run upgrade. So we just need the logic to compare it with the actual state in the cluster, ie 

```golang
if diff(actual, current) == nil {
  return
}
if diff(previous, current) == nil {
  rollback()
} else {
  upgrade()
}
```

I looked at https://github.com/databus23/helm-diff which seems to be doing exactly this but it doesn't look like the functionality can be easily imported. It also pulls in quite a lot of big deps for the actual diff calculation which we wouldn't need at this point.

**Comment by mkolesnik** [comment]
@dgn I saw this popped up again in a new bug, how about we dry run the upgrade before actually running it?
Then we can hash the generated manifest and save the hash in status.
Each time before we actually run the upgrade we can check if the hash changed or not, and apply it only if changed.

WDYT?
